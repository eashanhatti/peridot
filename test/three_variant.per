/* The natural number datatype. See `nat.per` */
define Nat
  : Type
  = Record {
      tag: Bool,
      data:
        if tag {
          Record {}
        } else {
          Nat
        }
    }

/* A `Bool` is not sufficient to describe a datatype with more than two variants, so here we use a `Nat` */
define Foo
  : Type
  = Record {
      tag: Nat,
      data:
        if tag.tag {
          Record { data: Nat }
        } elseif tag.data.tag {
          Record { a: Bool }
        } else {
          Record { data: Bool }
        }
    }
