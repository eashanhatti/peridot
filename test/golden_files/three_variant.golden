((fromList [], fromList [], fromList []),
 Declare
   Obj
   (Rigid
      (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
   (Rigid (TypeType Obj))
   (Declare
      Obj
      (Rigid
         (RNameIntro (UserName "Foo") Obj (Id {unId = 5})))
      (Rigid (TypeType Obj))
      (Define
         (Rigid
            (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
         (RecType
            (fromList
               [(Field {unField = "tag"}, Rigid TwoType),
                (Field {unField = "data"},
                 TwoElim
                   (LocalVar (Index {unIndex = 0}))
                   (RecType (fromList []))
                   (GlobalVar
                      (Rigid
                         (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
                      False))]))
         (Define
            (Rigid
               (RNameIntro (UserName "Foo") Obj (Id {unId = 5})))
            (RecType
               (fromList
                  [(Field {unField = "tag"},
                    GlobalVar
                      (Rigid
                         (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
                      False),
                   (Field {unField = "data"},
                    TwoElim
                      (RecElim
                         (LocalVar (Index {unIndex = 0}))
                         (Field {unField = "tag"}))
                      (RecType
                         (fromList
                            [(Field {unField = "data"},
                              GlobalVar
                                (Rigid
                                   (RNameIntro
                                      (UserName "Nat") Obj (Id {unId = 0})))
                                False)]))
                      (TwoElim
                         (RecElim
                            (RecElim
                               (LocalVar (Index {unIndex = 0}))
                               (Field {unField = "data"}))
                            (Field {unField = "tag"}))
                         (RecType
                            (fromList [(Field {unField = "a"}, Rigid TwoType)]))
                         (RecType
                            (fromList
                               [(Field {unField = "data"},
                                 Rigid TwoType)]))))]))
            (Declare
               Obj
               (Rigid
                  (RNameIntro (UserName "zero") Obj (Id {unId = 1})))
               (GlobalVar
                  (Rigid
                     (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
                  False)
               (Declare
                  Obj
                  (Rigid
                     (RNameIntro (UserName "succ") Obj (Id {unId = 2})))
                  (ObjFunType
                     Explicit
                     (GlobalVar
                        (Rigid
                           (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
                        False)
                     (GlobalVar
                        (Rigid
                           (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
                        False))
                  (Declare
                     Obj
                     (Rigid
                        (RNameIntro (UserName "add") Obj (Id {unId = 3})))
                     (ObjFunType
                        Explicit
                        (GlobalVar
                           (Rigid
                              (RNameIntro (UserName "Nat") Obj (Id {unId = 0})))
                           False)
                        (ObjFunType
                           Explicit
                           (GlobalVar
                              (Rigid
                                 (RNameIntro
                                    (UserName "Nat") Obj (Id {unId = 0})))
                              False)
                           (GlobalVar
                              (Rigid
                                 (RNameIntro
                                    (UserName "Nat") Obj (Id {unId = 0})))
                              False)))
                     (Define
                        (Rigid
                           (RNameIntro (UserName "zero") Obj (Id {unId = 1})))
                        (RecIntro
                           (fromList
                              [(Field {unField = "tag"}, Rigid TwoIntro0),
                               (Field {unField = "data"},
                                RecIntro (fromList []))]))
                        (Define
                           (Rigid
                              (RNameIntro
                                 (UserName "succ") Obj (Id {unId = 2})))
                           (ObjFunIntro
                              (RecIntro
                                 (fromList
                                    [(Field {unField = "tag"}, Rigid TwoIntro1),
                                     (Field {unField = "data"},
                                      LocalVar (Index {unIndex = 1}))])))
                           (Define
                              (Rigid
                                 (RNameIntro
                                    (UserName "add") Obj (Id {unId = 3})))
                              (ObjFunIntro
                                 (ObjFunIntro
                                    (TwoElim
                                       (RecElim
                                          (LocalVar (Index {unIndex = 1}))
                                          (Field {unField = "tag"}))
                                       (LocalVar (Index {unIndex = 0}))
                                       (ObjFunElim
                                          (ObjFunElim
                                             (GlobalVar
                                                (Rigid
                                                   (RNameIntro
                                                      (UserName "add")
                                                      Obj
                                                      (Id {unId = 3})))
                                                False)
                                             (RecElim
                                                (LocalVar (Index {unIndex = 1}))
                                                (Field {unField = "data"})))
                                          (LocalVar (Index {unIndex = 0}))))))
                              (Rigid (TypeType Obj))))))))))))
